wal: add write ahead log package

fabxc opened the pull request:
This adds a new WAL that's agnostic to the actual record contents.
It's much simpler and should be more resilient than the existing one.

It is safe for concurrent reads across different processes.

@brian-brazil @jkohen

Integrating this into the rest of TSDB is another can of worms. I may need to make some slight tweaks here and there, but in general it seems complete and working.

The batch mode is mostly there because I'd like to implement the checkpointing (formerly truncation) by just writing another WAL so we don't need to implement another disk format. But in that batch write scenario, firing off millions of tiny writes wouldn't be ideal.

Also we need some nicer recovery handling than in the last one – but should be simpler to do as well.

krasi-georgiev commented:
I smell some offline kubecon discussions :thinking: 
I am just guessing, but probably this is in relation to the WAL replication mentioned in @gouthamve's [talk](https://youtu.be/0UvKEHFNu4Q?t=1316). 

Are there any more discussions how and where would this be used?

krasi-georgiev commented:
I figured it to be in relation with this proposal
https://docs.google.com/document/d/1TEqqE_Stq04drhjSU1I7Ctmuy0dpsvlPL1AKxqEQoSg

krasi-georgiev commented on wal/wal.go:
```diff
@@ -0,0 +1,628 @@
+		default:
+			return errors.Errorf("unexpected record type %d", typ)
+		}
+		// Only increment i for non-zero records since we use it below
```
below -> above

fabxc commented on wal/wal.go:
```diff
@@ -0,0 +1,628 @@
+		if err := prev.Close(); err != nil {
+			level.Error(w.logger).Log("msg", "close previous segment", "err", err)
+		}
+	}
```
Unfortunately not, mutex would mean the writing goroutines would block. Fsync can take several seconds or much longer in bad conditions.

fabxc commented:
> the code is easy to follow and read 👍

Added 3 commits to change that :P

I know it's quite a bit of code for a single PR, but it's well-separated bottom-up along commits. I think some parts are just better to review with the full picture at hand.

Old WAL code was not removed. It needs to stick around anyway for migration procedures.

jkohen commented on wal/wal.go:
```diff
@@ -0,0 +1,815 @@
+
+// flushPage writes the contents of the page to disk.
+// If clear is set or no more records will reasonbly fit into it, its
+// remaining size gets written as zero data and its reset for the next physical page.
```
I'm not sure I understand the last part. Is it saying that the page will be padded with byte-0 if clear is true or the page doesn't have room for another record?

fabxc replied:
If `clear` is false, we may fill up the remainer of the page with zero bytes if we think there's not enough space left to fit in another (partial) record. If `clear` is true, we force the zero padding no matter how much space is left. We need the latter on shutdown or when completing a segment (when no partial records may be written), since we never want to leave a segment with a size that's not a multiple of 32KB.

I will make the comment clearer.

fabxc commented on wal/wal.go:
```diff
@@ -0,0 +1,815 @@
+func (w *WAL) fsync(f *Segment) error {
+	start := time.Now()
+	err := fileutil.Fsync(f.File)
+	w.fsyncDuration.Observe(time.Since(start).Seconds())
```
In Prometheus we've a few instrumentation conventions that are not baked into our model. One of them is that we always use base units, i.e. bytes, seconds, ...
The idea is that you don't have to make unit adjustments if you want to do binary ops between metrics and such and can just do a final division/multiplication on the overall result. Dashboard builders also often allow you to set a unit and they'll pick the most sensible scale – you could then just always default to the base unit and don't have to handpick it for each graph.

Since everything is a double in our world, this choice conveniently doesn't really impact data size or similar.

`Time.Seconds()` returns a float64 by the way, so we aren't losing the sub-seconds in case you meant that.

fabxc commented on checkpoint.go:
```diff
@@ -0,0 +1,268 @@
+		if err != nil || k >= n {
+			continue
+		}
+		if err := os.RemoveAll(filepath.Join(dir, fi.Name())); err != nil {
```
You are right. We shouldn't abort on failure in both places. Especially since those leftover checkpoints don't cause problems in general since we just ignore them.

fabxc commented:
Some initial feedback after running this in a small Prometheus server with high frequency scrapes:

Memory and CPU usage seem about 3-5% lower on average, which may just be noise. Max usage shows no notable difference.
The page flush to page completion ratio when scraping ~450 series is about 10x. So each 32KB WAL page gets written 10 times. Assuming a 4KB physical page size, we are looking at ~1.25 write amplification – so nothing to worry about there even if it had high variance.

Startups got about 3x slower. This is not super concerning but also not great for big setups and worth investigating.

fabxc commented:
@brian-brazil @gouthamve can I have a review from one of you?

--- 

We introduced a regression a while back regarding block order. It never made it into Prometheus though. Fixed it in here and added a test.

jkohen commented:
This is looking good.

fabxc commented:
During performance analysis of this change, I found a regression. A buffer that seems stack-allocated doesn't pass escape analysis and gets thrown on the heap. That resulted in big spikes. I pushed a commit to fix this.

This show memory usage before and after the change. It shows the released v2.2.1 and a version with the new WAL and in-memory metadata store. 

<img width="1602" alt="screen shot 2018-05-24 at 3 50 53 pm" src="https://user-images.githubusercontent.com/4948210/40510587-2751dfe8-5f6b-11e8-85ca-85ea91801b03.png">

brian-brazil commented on checkpoint.go:
```diff
@@ -0,0 +1,279 @@
+// This makes it easy to read it through the WAL package and concatenate
+// it with the original WAL.
+//
+// Non-critical errors are locked and not returned.
```
logged?

brian-brazil commented on checkpoint.go:
```diff
@@ -0,0 +1,279 @@
+			return nil, errors.New("invalid record type")
+		}
+		if len(buf[start:]) == 0 {
+			continue // all contents discarded
```
Comment style

brian-brazil commented on checkpoint.go:
```diff
@@ -0,0 +1,279 @@
+		}
+		recs = append(recs, buf[start:])
+
+		// flush records in 1 MB increments.
```
Capital letter

brian-brazil commented on db.go:
```diff
@@ -613,10 +616,6 @@ func OverlappingBlocks(bm []BlockMeta) Overlaps {
 	if len(bm) <= 1 {
 		return nil
 	}
-	sort.Slice(bm, func(i, j int) bool {
```
Why are you removing this?

fabxc replied:
This was mutating the input slice in a function that should do a read-only check. Thus current master actually has a regression. The sort is moved properly into the reload function.

fabxc replied:
This is fixed and tested in master via #335 so no longer relevant for this PR.

brian-brazil commented on fileutil/fileutil.go:
```diff
@@ -23,3 +24,25 @@ func ReadDir(dirpath string) ([]string, error) {
 	return names, nil
 }
+
+// Rename safely renames a file.
```
This has a race condition between the remove and the rename

brian-brazil replied:
The doc says it's "safe" however there's a race condition. In what way is the meant to be safe?

fabxc replied:
I meant to elaborate on what exactly the race condition is.

brian-brazil replied:
From the time the removeall starts to the rename ends, partial or no file may be present.

brian-brazil commented on head.go:
```diff
@@ -263,66 +275,130 @@ func (h *Head) ReadWAL() error {
+	return nil
+}
+
+// Init backfills data from the write ahead log and prepares the head for writes.
```
This is more a load than a backfill, which is confusing

brian-brazil commented on head.go:
```diff
@@ -348,18 +424,37 @@ func (h *Head) Truncate(mint int64) error {
+	if _, err = Checkpoint(h.logger, h.wal, m, n, keep, mint); err != nil {
+		return errors.Wrap(err, "create checkpoint")
 	}
 	h.metrics.walTruncateDuration.Observe(time.Since(start).Seconds())
```
We should probably have a separate metric for the checkpoint duration

fabxc replied:
It was renamed to checkpoint but as @jkohen pointed out, it might break alerts and dashboards. It is essentially still part of the truncation process, so keeping the name seemed fine. The `WAL.Truncate` call is just simple file removal. I don't think there's anything worth measuring in there.

brian-brazil commented on wal/wal.go:
```diff
@@ -0,0 +1,825 @@
+	w.segment = next
+	w.donePages = 0
+
+	// Don't block further writes by fsyinc the last segment.
```
fsyncing

brian-brazil commented on wal/wal.go:
```diff
@@ -0,0 +1,825 @@
+		}
+	}
+
+	donec := make(chan struct{})
```
I think a single-level channel is sufficient here

fabxc replied:
It's not, we have to wait until run terminated successfully. Currently it returns right when it reads from the channel. But this is quite implementation dependent and would be an easy regression target.

brian-brazil replied:
It looks like unnecessary complexity currently, especially as WAL is a struct rather than an interface.

brian-brazil commented on wal/wal.go:
```diff
@@ -0,0 +1,825 @@
+
+func (w *WAL) run() {
+	for {
+		// Processing all pending functions has precedence over shutdown.
```
What happens if we shutdown with pending functions? The current code seems to allow for that

fabxc commented on head.go:
```diff
@@ -221,17 +223,27 @@ func (h *Head) processWALSamples(
 			}
+			if s.T > maxt {
+				maxt = s.T
+			}
```
I actually uncovered several more issues around this general area. I've adjusted things to make it work, but it touches a few more places. So I'd rather add this in a PR on top.

krasi-georgiev commented:
btw did you try to run this with `--race` enabled ?

fabxc commented:
@krasi-georgiev yea, that looks all okay.

fabxc commented:
@brian-brazil how does this one look? I've two follow up PRs ready.

brian-brazil commented:
I'm still waiting on file format docs/protocol before giving it a full review.

fabxc commented:
The record encoding itself did not change. But so far the WAL format isn't documented at all. I've another PR in the queue that changes the record format and backfills documentation.

I can backport it to this one though without those changes.

fabxc commented:
@brian-brazil docs added

fabxc commented:
Yes, I just wanted to start getting this in since there are further changes on top of it and I wanted to enter a thorough benchmarking and validation phase sooner rather than later.

brian-brazil commented on head.go:
```diff
@@ -348,18 +424,37 @@ func (h *Head) Truncate(mint int64) error {
+		return nil // no segments yet.
+	}
+	// The lower third of segments should contain mostly obsolete samples.
+	// If we have too few segments, it's not worth checkpointing yet.
```
It'd be clearer if this said "If we don't have 3 segments"

fabxc commented:
> As an async reader of this file, how do I know that a page is fully flushed?

Pages are written partially, so we never write 0 bytes that we later fill with real data. A page is thus flushed, when a full page can be read.

gouthamve commented on checkpoint.go:
```diff
@@ -0,0 +1,279 @@
+			sr = last
+		}
+
+		segs, err := wal.NewSegmentsRangeReader(w.Dir(), m, n)
```
`segs` seems like an array of segments, `segsr` to indicate reader?

gouthamve commented on checkpoint.go:
```diff
@@ -0,0 +1,279 @@
+	stats := &CheckpointStats{}
+
+	var sr io.Reader
+	{
```
Why do we have a new block?

fabxc replied:
To not pollute the entire function with temporary variables we create here that we just need for a few lines.

gouthamve commented on checkpoint.go:
```diff
@@ -0,0 +1,279 @@
+			// Drop irrelevant tombstones in place.
+			repl := tstones[:0]
+			for _, s := range tstones {
+				for _, iv := range s.intervals {
```
Check for `keep(s.ref)` too?

fabxc replied:
Same as above. I'd err on the side of having a few MB of temporary overhead over calling something with unknown cost a billion times.

gouthamve commented on checkpoint.go:
```diff
@@ -0,0 +1,279 @@
+			// Drop irrelevant samples in place.
+			repl := samples[:0]
+			for _, s := range samples {
+				if s.T >= mint {
```
Check for `keep(s.Ref)` too?

fabxc replied:
That would help for data size. I wonder about the penalty though of calling a function for each sample – especially since we don't necessarily know how expensive that function is.

fabxc commented on docs/format/wal.md:
```diff
@@ -0,0 +1,72 @@
+
+Notable deviations are that the record fragment is encoded as:
+
+┌───────────┬──────────┬────────────┬──────────────┐
```
Oh, I believe we are just missing the code block formatting.

fabxc commented on head.go:
```diff
@@ -520,14 +627,41 @@ func (a *headAppender) AddFast(ref uint64, t int64, v float64) error {
+	var enc RecordEncoder
+
+	if len(a.series) > 0 {
+		rec = enc.Series(a.series, buf)
```
Mh, fair point – worth thinking about further. However I'd imagine rollbacks to be extremely rare. Combined with the small fraction of commits that actually create series, I wonder whether this can be an indicator for common issues.

fabxc replied:
Do we agree that this observation has nothing to do with this PR in particular and should be investigated/addressed seperately?

brian-brazil replied:
This PR is changing key code around this, and will likely affect the design of the fix. I'd prefer to see it in here - should be a matter of writing out series on a rollback too.

fabxc replied:
This is a big PR already. It doesn't change anything about the issue at hand, so I don't see why one would throw this on top here as well.
We can do another PR right after.

fabxc commented on wal/wal.go:
```diff
@@ -0,0 +1,828 @@
+func (r *Reader) next() (err error) {
+	// We have to use r.buf since allocating byte arrays here fails escape
+	// analysis and ends up on the heap, even though it seemingly should not.
+	hdr := r.buf[:7]
```
👍

fabxc commented on wal/wal.go:
```diff
@@ -0,0 +1,828 @@
+		f.Close()
+		return nil, err
+	}
+	// If the last page is torn, fill it with zeros.
```
That a write was torn, i.e. didn't complete entirely, due to a crash for example.

fabxc commented on wal/wal.go:
```diff
@@ -0,0 +1,828 @@
+
+// OpenWriteSegment opens segment k in dir. The returned segment is ready for new appends.
+func OpenWriteSegment(dir string, k int) (*Segment, error) {
+	// Only .active segments are allowed to be opened for write.
```
Obsolete comment.

fabxc commented:
> But for me the biggest question is that I am not seeing any versioning there. The segments have no headers, is that intentional?

Yea, I've been thinking about that a fair bit.
Ultimately I found per-segment versioning very weird. It would imply we may allow a mix of different versions, which we definitely don't want I'd say.
OTOH, the records we write to the WAL are still versioned – that's the most likely area of change. The record fragments the WAL implementation writes also have a flag byte that allows for versioning. That leaves the chance that we may want to migrate to yet another WAL, that cannot even keep the compatibility of having the first byte of a record (fragment) being a flag. That seemed extremely unlikely. And if, we can always add a version file into the WAL dir later that then at least applies to all segments equally. Doing this now seemed premature.

As for magic file headers – I think they are quite common but the practical usefulness seems rather low.

fabxc commented on docs/format/wal.md:
```diff
@@ -0,0 +1,72 @@
+e.g. `000000`, `000001`, `000002`, etc., and are limited to 128MB by default.
+A segment is written to in pages of 32KB. Only the last page of the most recent segment
+may be partial. A WAL record is an opaque byte slice that gets split up into sub-records
+should it exceed the remaining space of the current page. Records are never split across
```
Then we'll end up with a segment larger than the default size of 128MB. Will add that.

fabxc commented on checkpoint.go:
```diff
@@ -0,0 +1,279 @@
+		}
+	}
+
+	cpdir := filepath.Join(w.Dir(), fmt.Sprintf("checkpoint.%06d", n))
```
I bumped it. This is just the padding though, so even if it overflowed, it still won't wrap around.

brian-brazil replied:
You're depending on sorting these values as strings though, so could missort.

fabxc commented on wal/wal.go:
```diff
@@ -0,0 +1,828 @@
+type recType uint8
+
+const (
+	recPageTerm recType = 0 // rest of page is empty
```
I referenced the RocksDB docs which do a good job at explaining the whole approach. I see little value in duplicating it.
In the unlikely case they change their approach and docs, we can always change that.

brian-brazil replied:
It still needs to be documented here. "largely borrowed from" doesn't tell me the actual spec, which I expect to get entirely from reading this document standalone.

fabxc commented on wal/wal.go:
```diff
@@ -0,0 +1,828 @@
+		}
+	}
+	// Drain and process any remaining functions.
+	for {
```
It's not being closed anywhere so it wouldn't terminate. Changed this so the range works.

simonpasquier commented on docs/format/wal.md:
```diff
@@ -0,0 +1,80 @@
+# WAL Disk Format
+
+The write ahead log operates in segments that that are numbered and sequential,
```
"that ~~that~~ are numbered"

simonpasquier commented on wal/wal.go:
```diff
@@ -0,0 +1,822 @@
+
+// WAL is a write ahead log that stores records in segment files.
+// It must be read from start to end once before logging new data.
+// If an erroe occurs during read, the repair procedure must be called
```
s/erroe/error/

fabxc commented:
Rebased and resolved conflicts.

brian-brazil commented on checkpoint.go:
```diff
@@ -0,0 +1,279 @@
+	DroppedTombstones int
+	TotalSeries       int // Processed series including dropped ones.
+	TotalSamples      int // Processed samples inlcuding dropped ones.
+	TotalTombstones   int // Processed tombstones including droppes ones.
```
dropped

brian-brazil commented on docs/format/wal.md:
````diff
@@ -0,0 +1,88 @@
+└─────────────────────────────────────────────────────┘
+```
+
+[1][https://github.com/facebook/rocksdb/wiki/Write-Ahead-Log-File-Format]
````
GitHub isn't rendering this properly.

brian-brazil commented on wal/wal.go:
```diff
@@ -0,0 +1,822 @@
+
+const (
+	recPageTerm recType = 0 // rest of page is empty
+	recFull     recType = 1 // full record
```
Full stops and capital letters. There's others in this file too.

brian-brazil commented:
:+1:
